crypto_kem/ledakemlt32/leaktime/gf2x_arith_mod_xPplusOne.c

```diff
@@ -0,0 +1,584 @@
+        left_bit_shift_wide_n(2 * NUM_DIGITS_GF2X_ELEMENT, aux, sparse[0]);
+
+        for (unsigned int i = 1; i < nPos; i++) {
+            if (sparse[i] != INVALID_POS_VALUE) {
+                left_bit_shift_wide_n(2 * NUM_DIGITS_GF2X_ELEMENT, aux, (sparse[i] - sparse[i - 1]) );
+                PQCLEAN_LEDAKEMLT32_LEAKTIME_gf2x_add(resDouble, aux, resDouble, 2 * NUM_DIGITS_GF2X_ELEMENT);
+            }
+        }
+    }
+
+    gf2x_mod(Res, resDouble);
+
+}
+
+void PQCLEAN_LEDAKEMLT32_LEAKTIME_gf2x_transpose_in_place_sparse(int sizeA, POSITION_T A[]) {
+
+    POSITION_T t;
+    int i = 0, j;
+
+    if (A[i] == 0) {
+        i = 1;
+    }
+    j = i;
+
+    for (; i < sizeA && A[i] != INVALID_POS_VALUE; i++) {
+        A[i] = P - A[i];
+    }
+
+    for (i -= 1; j < i; j++, i--) {
+        t = A[j];
+        A[j] = A[i];
+        A[i] = t;
+    }
+
+}
+
+void PQCLEAN_LEDAKEMLT32_LEAKTIME_gf2x_mod_mul_sparse(size_t sizeR, POSITION_T Res[],
+        size_t sizeA, const POSITION_T A[],
+        size_t sizeB, const POSITION_T B[]) {
+
+    /* compute all the coefficients, filling invalid positions with P*/
+    size_t lastFilledPos = 0;
+    for (size_t i = 0 ; i < sizeA ; i++) {
+        for (size_t j = 0 ; j < sizeB ; j++) {
+            uint32_t prod = A[i] + B[j];
+            prod = ( (prod >= P) ? prod - P : prod);
+            if ((A[i] != INVALID_POS_VALUE) &&
+                    (B[j] != INVALID_POS_VALUE)) {
+                Res[lastFilledPos] = prod;
+            } else {
+                Res[lastFilledPos] = INVALID_POS_VALUE;
+            }
+            lastFilledPos++;
+        }
+    }
+    while (lastFilledPos < sizeR) {
+        Res[lastFilledPos] = INVALID_POS_VALUE;
+        lastFilledPos++;
+    }
+    PQCLEAN_LEDAKEMLT32_LEAKTIME_quicksort_sparse(Res);
+    /* eliminate duplicates */
+    POSITION_T lastReadPos = Res[0];
+    int duplicateCount;
+    size_t write_idx = 0;
+    size_t read_idx = 0;
+    while (read_idx < sizeR  && Res[read_idx] != INVALID_POS_VALUE) {
+        lastReadPos = Res[read_idx];
+        read_idx++;
+        duplicateCount = 1;
+        while ( (Res[read_idx] == lastReadPos) && (Res[read_idx] != INVALID_POS_VALUE)) {
+            read_idx++;
+            duplicateCount++;
+        }
+        if (duplicateCount % 2) {
+            Res[write_idx] = lastReadPos;
+            write_idx++;
+        }
+    }
+    /* fill remaining cells with INVALID_POS_VALUE */
+    for (; write_idx < sizeR; write_idx++) {
+        Res[write_idx] = INVALID_POS_VALUE;
+    }
+}
+
+/* the implementation is safe even in case A or B alias with the result */
+/* PRE: A and B should be sorted and have INVALID_POS_VALUE at the end */
+void PQCLEAN_LEDAKEMLT32_LEAKTIME_gf2x_mod_add_sparse(
+    int sizeR, POSITION_T Res[],
+    int sizeA, const POSITION_T A[],
+    int sizeB, const POSITION_T B[]) {
+
+    POSITION_T tmpRes[DV * M]; // TODO: now function only works for adding (disjunct) DV and M positions
```

Comment:
how about this TODO?

Reply:
It's only used for adding disjunct locations. I'll add it as a precondition.